app/Platform/Listeners/DispatchUpdatePlayerRanksJob.php

```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace App\Platform\Listeners;
+
+use App\Platform\Events\PlayerGameMetricsUpdated;
+use App\Platform\Events\PlayerRankedStatusChanged;
+use App\Platform\Jobs\UpdatePlayerRanksJob;
+use App\Site\Models\User;
+use Illuminate\Contracts\Queue\ShouldQueue;
+
+class DispatchUpdatePlayerRanksJob implements ShouldQueue
+{
+    public function handle(object $event): void
+    {
+        $user = null;
+
+        switch ($event::class) {
+            case PlayerGameMetricsUpdated::class:
```

Comment:
it looks like you're listening for `PlayerBadgeAwarded` and `PlayerBadgeLost`, not `PlayerGameMetricsUpdated`

Reply:
Want to confirm if this still needs to change now that `UpdatePlayerRanks` is using `SiteAwards` as its source of truth for aggregates:

https://github.com/RetroAchievements/RAWeb/pull/1999/files#diff-d1cd771d569c8de22cd2a5fbb203f1679cdd74af5e55f288f169697ed8bbeff8R28

Reply:
I'm referring to this code: https://github.com/RetroAchievements/RAWeb/blob/3775e153fab8d826fe1005afc595d87ae2ebdde1/app/Platform/EventServiceProvider.php#L74-L81

This listener will get called with a `PlayerBadgeAwarded` event or a `PlayerBadgeLost` event. I don't see anywhere where it would get called with a `PlayerGameMetricsUpdated` event.

Reply:
Ah, I think I understand now. I've removed the `DispatchUpdatePlayerStatsJob` from these two in latest and moved it over to `PlayerGameMetricsUpdated`.

Reply:
If the queries rely strictly on the `SiteAwards` table, wouldn't it be more succinct to have the listener watching for `PlayerBadgeAwarded` and `PlayerBadgeLost`?

`UpdatePlayerGameMetrics` disptches a `RevalidateAchievementsSetBadgeEligibility` action, which would raise the `PlayerBadgeAwarded` or `PlayerBadgeLost` when appropriate, and you wouldn't be regenerating the stats any time the player unlocked an achievement.

Reply:
Sorry, I was getting confused about multiple things simultaneously, but I think I understand the feedback now.

I've updated `EventServiceProvider.php` so it's listening for `PlayerBadgeAwarded` and `PlayerBadgeLost` again (and not `PlayerGameMetricsUpdated`). I've also updated `DispatchUpdatePlayerStatsJob` to have cases for `PlayerBadgeAwarded` and `PlayerBadgeLost` events.

Reply:
Latest commit https://github.com/RetroAchievements/RAWeb/pull/1999/commits/763b0405633a16800ff7003a0eb9887bd6cc9980 resolves an issue where `PlayerBadgeLost` would instantly fail due to a race condition from the badge itself no longer existing.

`PlayerBadgeLost` is now dispatched with a deconstructed badge instance:

```php
private function dispatchBadgeLostEvent(PlayerBadge $badge): void
{
    PlayerBadgeLost::dispatch(
        $badge->user,
        $badge->AwardType,
        $badge->AwardData,
        $badge->AwardDataExtra,
    );
}
```